platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/EncryptingPersistenceManager.java

```diff
@@ -0,0 +1,208 @@
+        Dictionary<String, Object> encryptedDictionaryToReturn = new Hashtable<>();
+        Enumeration<String> keys = dict.keys();
+        while (keys.hasMoreElements()) {
+            String key = keys.nextElement();
+            encryptedDictionaryToReturn.put(key, encryptIfNecessary(key, dict.get(key)));
+        }
+        return encryptedDictionaryToReturn;
+    }
+
+    // The 2.3.7 version of the maven-bundle-plugin fails on lambdas,
+    // so we need to use Java 7 constructs
+    private Dictionary<String, Object> decryptAndReturnDictonary(Dictionary<String, Object> dict) {
+        Dictionary<String, Object> decryptedDictionaryToReturn = new Hashtable<>();
+        Enumeration<String> keys = dict.keys();
+        while (keys.hasMoreElements()) {
+            String key = keys.nextElement();
+            decryptedDictionaryToReturn.put(key, decryptIfNecessary(key, dict.get(key)));
+        }
+        return decryptedDictionaryToReturn;
+    }
+
+    private Object encryptIfNecessary(String key, Object value) {
+        if (value instanceof String && !EXCLUDED_PROPERTIES.contains(key)) {
+            return agent.encrypt((String) value);
+        }
+        return value;
+    }
+
+    private Object decryptIfNecessary(String key, Object value) {
+        if (value instanceof String && !EXCLUDED_PROPERTIES.contains(key)) {
+            return agent.decrypt((String) value);
+        }
+        return value;
+    }
+
+    /**
+     * PAX Logging might not be fully configured and ready if there is a failure in the agent's
+     * constructor. To remedy this, we will cache failure data to later throw back.
+     */
+    private class EncryptionAgent {
+        private String initFailureMessage = null;
+
+        private Exception initException = null;
+
+        private Crypter crypter;
+
+        EncryptionAgent(String pwDirString) {
+            File pwDir = new File(pwDirString);
+            if (!pwDir.mkdirs() && !pwDir.exists()) {
+                initFailureMessage = "Cannot create required directory for encryption ["
+                        + pwDir.getAbsolutePath() + "]. This might be a permissions issue. Check "
+                        + "OS settings, file permissions, and the active user account. Then unzip a "
+                        + "new instance and try again";
+                throw new IllegalStateException(initFailureMessage);
+            }
+
+            String[] children = pwDir.list();
+            if (children == null) {
+                initFailureMessage = "Expected directory [" + pwDir.getAbsolutePath() + "] to "
+                        + "exist and be accessible. Check OS settings, file permissions, and the "
+                        + "active user account. Then unzip a new instance and try again";
+                throw new IllegalStateException(initFailureMessage);
+            }
+
+            synchronized (EncryptionAgent.class) {
```

Comment:
❓ Curious as to why we need to synchronize on the class here?

Reply:
See below comment.